keel-tagging-plugin/src/main/kotlin/com/netflix/spinnaker/keel/tagging/ResourceTagger.kt

```diff
@@ -0,0 +1,203 @@
+/*
+ *
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.netflix.spinnaker.keel.tagging
+
+import com.netflix.spinnaker.keel.activation.ApplicationDown
+import com.netflix.spinnaker.keel.activation.ApplicationUp
+import com.netflix.spinnaker.keel.actuation.ResourcePersister
+import com.netflix.spinnaker.keel.api.ResourceName
+import com.netflix.spinnaker.keel.api.SPINNAKER_API_V1
+import com.netflix.spinnaker.keel.api.SubmittedResource
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.clouddriver.model.Credential
+import com.netflix.spinnaker.keel.events.DeleteEvent
+import com.netflix.spinnaker.keel.persistence.NoSuchResourceException
+import com.netflix.spinnaker.keel.persistence.ResourceRepository
+import com.netflix.spinnaker.keel.tags.EntityRef
+import com.netflix.spinnaker.keel.tags.EntityTag
+import com.netflix.spinnaker.keel.tags.KEEL_TAG_NAME
+import com.netflix.spinnaker.keel.tags.TagValue
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.scheduling.annotation.Scheduled
+import java.time.LocalDateTime
+import java.time.ZoneOffset
+import java.util.concurrent.TimeUnit
+
+/**
+ * A scheduled job that checks to make sure each resource keel is managing
+ * has an entity tag indicating this, as well as linking the entity tag id
+ * to the keel id.
```

Comment:
If we do this, I'm not clear why we need to model tags as their own resource type. Shouldn't we _either_ periodically check all our resources for tags, _or_ do something after a resource is created or deleted?

Reply:
As we discussed offline, I'm going to switch to doing something on create and delete + daily removing old "tag not desired" states. 

We can revisit in the future if there's a better optimization.

Re: clouddriver entity tags performance: currently you need elastic search for entity tags, and the correct way is going through clouddriver. Adam does not think there's a performance gain to doing a bulk update of the tags.